### ISSUE TITLE
android GpuDelegate: even the simplest tflite converted models seems to run on CPU

### ISSUE THREAD

**Issue by bioothod** [tensorflow/tensorflow#30313]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): tf docker images
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: oneplus5, pixel3, galaxy s10
- TensorFlow installed from (source or binary): various
- TensorFlow version (use command below): 1.12.0, 1.13.1, 1.14.0, 1.15.0-dev20190628, r1.14 source build
- Python version: 3.x
- Bazel version (if compiling from source): 0.25.2
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
tflite converted models seems to always run on cpu despite using gpudelegate, for example the simplest model from https://github.com/tensorflow/tensorflow/issues/30311 runs for hundreds of milliseconds per inference no matter whether cpu or gpudelegate is used, while stock mobilenet_v1 (no quantization, no manual tflite conversion, downloaded tflite model) runs ~80ms per inference on this device when using gpudelegate and 150-300ms on cpu.

**Describe the expected behavior**
tflite to actually use GPU when running conv2d networks

**Code to reproduce the issue**
attached NHW3 model (and script to generate pb and tflite with optimization options, which do not have any effect), this is basically NHW3 input, single tf.layers.conv2d() constant-initialized layer and reduce_sum() for output.

Any hint on debugging this issue?
I tested both jcenter downloaded (0.0.0-nightly branches) and manually built from r1.14 branch tensorflow-lite and tensorflow-lite-gpu.
I used 1.12.0, 1.13.1, 1.14.0, 1.15.0-dev20190628, r1.14 source build tensorflow/tflite with different optimization options with no success.

Could you please show how *exactly* you generated mobilenet_v1 tflite model found in your tutorials, since it works like a charm with gpudelegate: ~80ms per single-image inference on gpu and several hundreds seconds on cpu on sufficiently recent android. Or any hints on how to debug this gpudelegate issue.

[test_model_3channels.tar.gz](https://github.com/tensorflow/tensorflow/files/3350552/test_model_3channels.tar.gz)

**Comment by impjdi** [comment]
@bioothod 

Thanks for attaching the model.

> ERROR: Next operations are not supported by GPU delegate:
> SUM: Operation is not supported.
> First 1 operations will run on the GPU, and the remaining 1 on the CPU.

Well, we don't have a SUM operation, and MobileNet v1 doesn't have a SUM at the end either.  It's not clear where the confusion is coming from.  Can you elaborate?

**Comment by impjdi** [comment]
@bioothod 

Oh and a couple of things.

1. By default, TFLite gpu delegate options has allow_precision_loss to 0 (false), to be on the safe side.  You may want to turn that on for speed up.
2. Your network needs to be deep enough.  There is additional cost you have to eat up (memcpy, GPU/CPU sync etc.) that needs to be compensated by GPU's faster execution.  So if you have a network with just a single CONV_2D, of course it's going to be slower than on the CPU only.

**Comment by bioothod** [comment]
https://www.tensorflow.org/lite/guide/ops_compatibility states that tf.reduce_sum will be removed from the graph, probably will be replaced with ADD_N operation. Or does it mean it should be removed by the user? I've attached different mobilenet_v1 which doesn't work as fast as google-converted tflite either, it doesn't have SUM operator.

Also. I'm quite sure I saw gpudelegate exception about unsupported SUM operator with some other models (do not remember, what it was though), and now there are none, model just works slow.

I used precision loss without any effect.
This conv2d network is slower than the whole mobilenet_v1 on cpu, well, its because of large convolutions, but still.

[model.ckpt-1044627-frozen.tflite.gz](https://github.com/tensorflow/tensorflow/files/3352660/model.ckpt-1044627-frozen.tflite.gz)

**Comment by bioothod** [comment]
N.B. maybe not slower, after 100-400 ms variation is very high and I never measured mean and anyway it is unacceptably high for 2 operations or for the whole network like in the example above. And since there is no exception in gpudelegate log I suppose it runs on GPU and should be, well, not slower than google-converted mobilenet_v1

**Comment by bioothod** [comment]
More confusion.

SUM operator works and produces correct results (maybe running on CPU - any chance to **know** that graph works on GPU?):
```
images = tf.placeholder(tf.float32, shape=[None, 32, 32, num_channels], name='input/tensor')
x = tf.layers.conv2d(images, 1, kernel_size=[3, 3], kernel_initializer=tf.ones_initializer(), use_bias=False)
sum_pixels = tf.reduce_sum(x, axis=[1, 2], name='output/sum_pixels')
```
[test_conv2d_3channels_sum_works.tflite.gz](https://github.com/tensorflow/tensorflow/files/3356009/test_conv2d_3channels_sum_works.tflite.gz)


SUM operator crashes with unsupported op exception:
```
images = tf.placeholder(tf.float32, shape=[None, 32, 32, num_channels], name='input/tensor')
x = tf.layers.conv2d(images, 1, kernel_size=[32, 32], kernel_initializer=tf.ones_initializer(), use_bias=False)
sum_pixels = tf.reduce_sum(x, axis=[1, 2], name='output/sum_pixels')
```
[test_conv2d_3channels_sum_crashes.tflite.gz](https://github.com/tensorflow/tensorflow/files/3355986/test_conv2d_3channels_sum_crashes.tflite.gz)


GpuDelegate crashes with:
```
I/Adreno: Error: Uniform offsets cannot find a suitable location/component.
    Error: Linking failed.
W/ImageReader_JNI: Unable to acquire a buffer item, very likely client tried to acquire more than maxImages buffers
E/AndroidRuntime: FATAL EXCEPTION: inference
    Process: org.tensorflow.lite.examples.classification, PID: 30676
    java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: TfLiteGpuDelegate Prepare: Program is not properly linked: Error: Uniform offsets cannot find a suitable location/component.
    Error: Linking failed.Node number 2 (TfLiteGpuDelegate) failed to prepare.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.applyDelegate(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.init(NativeInterpreterWrapper.java:83)
        at org.tensorflow.lite.NativeInterpreterWrapper.<init>(NativeInterpreterWrapper.java:60)
        at org.tensorflow.lite.Interpreter.<init>(Interpreter.java:224)
```
```
images = tf.placeholder(tf.float32, shape=[None, 32, 32, num_channels], name='input/tensor')
x = tf.layers.conv2d(images, 1, kernel_size=[32, 32], kernel_initializer=tf.ones_initializer(), use_bias=False)
sum_pixels = tf.reshape(x, [1, 1], name='output/sum_pixels')
```
[test_conv2d_3channels_crash_uniform_offsets.tflite.gz](https://github.com/tensorflow/tensorflow/files/3356051/test_conv2d_3channels_crash_uniform_offsets.tflite.gz)

**Comment by impjdi** [comment]
`delegates/gpu/common/model_builder.cc` has a function called `NewOperationParser()`.  That's the set of operations supported by GPU.

**Comment by bioothod** [comment]
> `delegates/gpu/common/model_builder.cc` has a function called `NewOperationParser()`. That's the set of operations supported by GPU.

It is probably related to SUM operation. Well, then why does it work in the first graph? The only difference is input shape [1,1,1,1](crashes) vs [1,1,1,1](works)

**Comment by impjdi** [comment]
In all cases, SUM runs on the CPU.

What is succeeding on the first case and failing on the second case is CONV_2D on the GPU.

**Comment by bioothod** [comment]
Things are getting more and more weird, isn't conv_2d supposed to always succeed, at least as it is written in documentation, or there are some constraints? Why doesn't delegate throw the same exception explaining (at least something)...

Are you on this issue and do you need any additional details from me, or you think everything is fine?
I will dig into delegate code in a meantime (ouch, there are no debug helpers) and try to find out more

**Comment by bioothod** [comment]
Apparently, both failing models fail because GL can not compile conv_2d kernel for 32x32 reception field, which is very large for real life anyway, and for smaller kernels it silently replaces sum or squeeze wtih CPU implementations, just like you've said.

Still curious how to determine which operations run on gpu, since the only way to run fast enough is not to touch any property of tflite converter, even not `setting converter.optimizations = [tf.lite.Optimize.DEFAULT]`. Model runs fast with plain
```
converter = tf.lite.TFLiteConverter.from_session(sess, [images], output_nodes)
#converter.optimizations = [tf.lite.Optimize.DEFAULT]
#converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_LATENCY]
#converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_SIZE]
converter = tf.lite.TFLiteConverter.from_session(sess, [images], output_nodes)
```

If any optimization property has been touched to any commented above, model greatly shrinks in size, but starts executing only on CPU.

**Comment by impjdi** [comment]
Again... 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/common/model_builder.cc#L1879-L1957

this is the list of supported ops on the GPU.

If it crashes, that's a bug, and we should look into it, but we cannot accommodate all hypothetical single op test cases.  After all, this is an open source project, and you can contribute by fixing bugs too ;)

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
